Claw state machine and states

katiewang22 opened the pull request.

katiewang22 committed:
claw states and configs

katiewang22 committed:
open close requests

ChristianPawlowicz commented on src/main/java/frc/robot/lib/states/Claw.java:
```diff
@@ -0,0 +1,34 @@
+    double closeConePosition = 0.0;
+
+    public Claw() {
+        RobotMap.claw.setNeutralMode(NeutralMode.Brake);
```
The claw variable needs to be created in RobotMap. It should also be more descriptive that just "claw". It is a motor class that commands the motor to open and close the claw. The name should reflect that.

ChristianPawlowicz commented on src/main/java/frc/robot/lib/states/Claw.java:
```diff
@@ -0,0 +1,34 @@
+public class Claw {
+
+    // TODO: positions
+    double closeCubePosition = 0.0;
```
Constants like this should be of the `SCREAMING_SNAKE_CASE` format. We should start enforcing programming [naming conventions](https://www.pluralsight.com/blog/software-development/programming-naming-conventions-explained#:~:text=Screaming%20Snake%20Case,-Snek%3A%20IT'S%20A&text=SCREAMING_SNAKE_CASE%20is%20a%20variable%20naming,C%2C%20Javascript%2C%20and%20Java.) for all our variables. All constants should be defined in a singular location in our code base so that we don't have to go hunting for them.

ChristianPawlowicz commented on src/main/java/frc/robot/subsystems/claw/ClawStateMachine.java:
```diff
@@ -0,0 +1,23 @@
+/** Add your docs here. */
+public class ClawStateMachine extends StateMachine {
+    
+    public static OpenState openState = new OpenState();
```
When we were discussing how the [claw state machine](https://lucid.app/lucidchart/b6090843-0b3b-4749-8b86-71403e67fac1/edit?invitationId=inv_7482c25f-2845-4fb0-9f3d-8a34f9c4208e&referringApp=slack&page=0_0#) works, we also included 'closing' and 'opening' states. These are not reflected in here.

ChristianPawlowicz commented on src/main/java/frc/robot/subsystems/claw/ClawStateMachine.java:
```diff
@@ -0,0 +1,23 @@
+        closedCone.build();
+        closedCube.build();
+
+        setCurrentState(openState);
```
Just a note here, we might not want to start in this state. When we start a match, we will ideally have some kind of pre-load. Not sure how we should handle that. This is fine for now.

ChristianPawlowicz commented on src/main/java/frc/robot/subsystems/claw/ClosedCone.java:
```diff
@@ -0,0 +1,37 @@
+    @Override
+    public void build() {
+        // open claw if open request
+        transitions.add(new Transition(() -> {
```
We should also add a transition if we don't detect a cone, go to the next state. Update state machine [diagram](https://lucid.app/lucidchart/b6090843-0b3b-4749-8b86-71403e67fac1/edit?invitationId=inv_7482c25f-2845-4fb0-9f3d-8a34f9c4208e&referringApp=slack&page=0_0#) to reflect this.

ChristianPawlowicz commented on src/main/java/frc/robot/subsystems/claw/ClosedCone.java:
```diff
@@ -0,0 +1,37 @@
+
+    @Override
+    public void execute() {
+        RobotMap.clawLib.setClawClosedCone();
```
We do not need to execute this multiple times. The command to set the motor to a position only needs to occurr once.

ChristianPawlowicz commented on src/main/java/frc/robot/subsystems/claw/ClosedCube.java:
```diff
@@ -0,0 +1,37 @@
+
+    @Override
+    public void execute() {
+        RobotMap.clawLib.setClawClosedCube();
```
We do not need to execute this multiple times. The command to set the motor to a position only needs to occurr once.

ChristianPawlowicz commented on src/main/java/frc/robot/subsystems/claw/OpenState.java:
```diff
@@ -0,0 +1,43 @@
+
+    @Override
+    public void execute() {
+        RobotMap.clawLib.setClawOpen();
```
We do not need to execute this multiple times. The command to set the motor to a position only needs to occurr once.

ChristianPawlowicz commented on src/main/java/frc/robot/subsystems/claw/OpenState.java:
```diff
@@ -0,0 +1,43 @@
+    public void build() {
+        // close on a cube if "close request" and color sensor == purple
+        transitions.add(new Transition(() -> {
+            return RobotMap.closeRequest && RobotMap.arm.getColorSensor() == "purple";
```
The color sensor should return some kind of enumerated value, not a string.

ChristianPawlowicz reviewed (changes requested):
Please rename this pull request to something reflecting the changes being made. These are not to-do lists.

katiewang22 committed:
claw color sensor and states updated

katiewang22 committed:
claw states updated

katiewang22 committed:
match confidence condition

uni-mairata committed:
Open claw button for all scoring states

uni-mairata committed:
delete openRequest for automatically opening claw

same reason as why it was deleted in final-arm-encoders (remove the automatic portion of opening claw, just use a button to account for if arm encoders are wrong)

katiewang22 committed:
move constants to Constants.java

katiewang22 committed:
Merge branch 'main' into final-claw

katiewang22 committed:
Fixed merge conflicts
